LWT: enable for tablet-based tables

gusev-p opened the pull request:
This PR enables **LWT (Lightweight Transactions)** support for tablet-based tables by leveraging **colocated tables**.

### Problem
Currently, storing Paxos state in system tables causes two major issues:
* **Loss of Paxos state during tablet migration or base table rebuilds**
  * When a tablet is migrated or the base table is rebuilt, system tables don't retain Paxos state.
  * This breaks LWT correctness in certain scenarios.
  * Failing test cases demonstrating this:
      * test_lwt_state_is_preserved_on_tablet_migration
      * test_lwt_state_is_preserved_on_rebuild
* **Shard misalignment and performance overhead**
  * Tablets may be placed on arbitrary shards by the tablet balancer.
  * Accessing Paxos state in system tables could require a shard jump, degrading performance.

### Solution
We move Paxos state into a dedicated Paxos table, colocated with the base table:
  * Each base table gets its own Paxos state table.
  * This table is lazily created on the first LWT operation.
  * Its tablets are colocated with those of the base table, ensuring:
    * Co-migration during tablet movement
    * Co-rebuilding with the base table
    * Shard alignment for local access to Paxos state

Some reasoning for why this is sufficient to preserve LWT correctness is discussed in [2].

This PR addresses two issues from the "Why doesn't it work for tablets" section  in [1]:
  * Tablet migration vs LWT correctness
  * Paxos table sharding

Other issues ("bounce to shard" and "locking for intranode_migration") have already been resolved in previous PRs.

References
[1] - [LWT over tablets design](https://docs.google.com/document/d/1CPm0N9XFUcZ8zILpTkfP5O4EtlwGsXg_TU4-1m7dTuM/edit?tab=t.0#heading=h.goufx7gx24yu)
[2] - [LWT: Paxos state and tablet balancer](https://docs.google.com/document/d/1-xubDo612GGgguc0khCj5ukmMGgLGCLWLIeG6GtHTY4/edit?tab=t.0)
[3] - [Colocated tables PR](https://github.com/scylladb/scylladb/pull/22906#issuecomment-3027123886)
[4] - [Possible LWT consistency violations after a topology change](https://github.com/scylladb/scylladb/issues/5251)

Backport: not needed because this is a new feature.

avikivity commented on service/paxos/paxos_state.cc:
```diff
@@ -366,11 +368,104 @@ static sstring paxos_state_cf_filter(const schema& s, const schema& state_schema
 
-paxos_store::paxos_store(db::system_keyspace& sys_ks)
-: _sys_ks(sys_ks)
+static const sstring paxos_state_table_suffix = "_scylla_paxos_state";
```
Let's use "$paxos" to make it shorter and to use the character traditionally designated for internal stuff.

gusev-p replied:
Do you mean as a prefix `$paxos_<base_table_name>` or suffix is ok?

avikivity replied:
I means base$paxos, it's nice to get the names grouped when sorted.

gusev-p replied:
OK, will update.

avikivity commented on service/paxos/paxos_state.cc:
```diff
@@ -366,11 +368,104 @@ static sstring paxos_state_cf_filter(const schema& s, const schema& state_schema
+static schema_ptr create_paxos_state_schema(const schema& s, std::optional<table_id> table_id) {
+    schema_builder builder(table_id, s.ks_name(), paxos_state_table_name(s),
+        // partition key
+        {{"row_key", bytes_type}}, // byte representation of a row key that hashes to the same token as original
```
We should rename to partition_key since we have the opportunity.

avikivity replied:
Ok

avikivity commented on service/paxos/paxos_state.cc:
```diff
@@ -366,11 +368,104 @@ static sstring paxos_state_cf_filter(const schema& s, const schema& state_schema
+    );
+    builder.set_gc_grace_seconds(0);
+    builder.with_hash_version();
+    return builder.build(schema_builder::compact_storage::no);
```
But, this table should not be repaired or replicated, right?

gusev-p replied:
It's never replicated. It may be repaired in case of base table rebuild or repair.

avikivity replied:
Need to prevent read-repair etc.

I guess it's not replicated because you don't use coordinator APIs. I think it's okay.

avikivity replied:
Makes sense. Filtering is suboptimal compared to preparing a replica set of locals in advance (in effective_replication_map) but we can optimize later.

avikivity commented on service/paxos/paxos_state.cc:
```diff
@@ -366,11 +368,104 @@ static sstring paxos_state_cf_filter(const schema& s, const schema& state_schema
+static schema_ptr create_paxos_state_schema(const schema& s, std::optional<table_id> table_id) {
+    schema_builder builder(table_id, s.ks_name(), paxos_state_table_name(s),
+        // partition key
+        {{"row_key", bytes_type}}, // byte representation of a row key that hashes to the same token as original
```
Agree

scylladb-promoter commented:
## Docs Preview :book:
Docs Preview for this pull request is available [here](http://previews.docs.scylladb.com/scylladb/24819)

**Changed Files:**
- [alternator/new-apis](http://previews.docs.scylladb.com/scylladb/24819/alternator/new-apis)

**Note:** This preview will be available for 30 days and will be automatically deleted after that period. You can manually trigger a new build by committing changes.

dorlaor commented:
Hallelujah!

On Sat, Jul 5, 2025 at 3:22 PM scylladb-promoter ***@***.***>
wrote:

> *scylladb-promoter* left a comment (scylladb/scylladb#24819)
> <https://github.com/scylladb/scylladb/pull/24819#issuecomment-3038864055>
> 🟢 CI State: SUCCESS
> Mode Stage Status Node
> Framework test
> <https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18138/> ✅
> i-055bd55aaa01effeb
> dev Build
> <https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18138/> ✅
> spider5.cloudius-systems.com
> test.py
> <https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18138/> ✅
> spider5.cloudius-systems.com
> Unit Tests Custom
> <https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18138/>
> 🔹 alternator
> 🔹 alternator/test_ttl
> 🔹 cluster/test_tablets2 ✅ spider5.cloudius-systems.com
> release Build
> <https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18138/> ✅
> spider7.cloudius-systems.com
> test.py
> <https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18138/> ✅
> spider7.cloudius-systems.com
> Unit Tests Custom
> <https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18138/>
> 🔹 alternator
> 🔹 alternator/test_ttl
> 🔹 cluster/test_tablets2 ✅ spider7.cloudius-systems.com
> debug Build
> <https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18138/> ✅
> spider4.cloudius-systems.com
> test.py
> <https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18138/> ✅
> spider4.cloudius-systems.com
> Unit Tests Custom
> <https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18138/>
> 🔹 alternator
> 🔹 alternator/test_ttl
> 🔹 cluster/test_tablets2 ✅ spider4.cloudius-systems.com
> dtest dtest with consistent topology changes
> <https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/9145/>
> ✅
> dtest with gossip topology changes
> <https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/17237/>
> ✅
> dtest with tablets
> <https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5918/>
> ✅ Build Details:
>
>    - Duration: 20 hr
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/scylladb/scylladb/pull/24819#issuecomment-3038864055>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AANHURPVTHFT2FSXC3BQZIL3G67R5AVCNFSM6AAAAACAWSCRE6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTAMZYHA3DIMBVGU>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

nyh commented on test/alternator/test_tablets.py:
```diff
@@ -118,14 +113,9 @@ def test_alternator_tablets_and_lwt(dynamodb):
 
-# An Alternator table created tablets and with a write isolation
-# mode that doesn't use LWT ("forbid_rmw") works normally, even
-# before #18068 is fixed.
```
There was no reason to remove this comment, which explains well why this test exists...

gusev-p replied:
I honestly can't imagine a situation where this comment adds value. The information it conveys can be deduced from the test name and the use of 'forbid_rmw'. However, this is not the point of this commit, I'll return this comment back.

nyh reviewed (changes requested):
Looks very good. It's a very long series (nice work!), so I did have a not-very-small amount of comments and request, but I think none of my comments are really serious or hard to fix (some of them are probably just my misunderstandings and won't require fixing anything). Also because this series is so long, in some cases I know I wrote comments on an early patch that I saw resolved in some later patch. In such cases, sorry about the spam.

I'm a bit afraid that I might have missed the forest for all the trees - reviewing many tiny details and not reviewing the real "LWT" picture well, so let's get somebody who's more of an expert on LWT to also review this PR.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -132,6 +132,14 @@ void validate_read_replicas(const locator::effective_replication_map& erm, const
+    auto it = std::ranges::remove_if(host_ids, [&my_id](locator::host_id& id) { 
+        return id != my_id;
+    }).begin();
+    host_ids.erase(it, host_ids.end());
```
Ah, I see later you give the same treatment to natural_endpoints and pending_endpoints.

gusev-p commented on service/paxos/paxos_state.cc:
```diff
@@ -366,11 +368,104 @@ static sstring paxos_state_cf_filter(const schema& s, const schema& state_schema
+
+static sstring paxos_state_table_name(const schema& s) {
+    return sstring(s.cf_name()) + paxos_state_table_suffix;
+}
```
I don't think we should put too many guardrails around these tables. Having CQL access to them is useful for debugging/troubleshooting.

Actually, the same can be said about various table parameters, like e.g. compaction strategy. If something goes wrong it's better to have all the standard tools available to fix the problem. This means the information about the table should be available to the user, including `DESCRIBE SCHEMA` output.

What if we just describe these tables in the documentation and won't try to hide them at all?  We'll mention that a paxos state table is created lazily when the first LWT is run against the base table, this will make it clear why we have a large latency for the first LWT.

@avikivity @gleb-cloudius @nyh

gleb-cloudius replied:
What about deleting them though? Do we allow to delete CDC tables? Also since we do not have a way to write into individual replicas yet writing to the table from CQL will unlikely behave in a way we want it to.

gusev-p replied:
> Do we allow to delete CDC tables

@piodul ?

> What about deleting them though?

This would be the user shooting themselves in the foot, but nothing catastrophic -- the `$paxos` table will be recreated on next LWT, only unfinished paxos rounds will be lost.

> Also since we do not have a way to write into individual replicas yet writing to the table from CQL will unlikely behave in a way we want it to.

Yes, but `replica_local` flag introduced in this PR makes it easy to implement `LOCAL_ONLY` CQL extension, proposed in [this ticket ](https://github.com/scylladb/scylladb/issues/16478). I can do this in a follow-up PR.

gleb-cloudius commented on service/storage_proxy.cc:
```diff
@@ -6484,8 +6484,11 @@ future<bool> storage_proxy::cas(schema_ptr schema, cas_shard cas_shard, shared_p
+            auto msg = format("Cannot use LightWeight Transactions for table {}.{}: LWT is not yet supported with tablets", schema->ks_name(), schema->cf_name());
+            co_await coroutine::return_exception(exceptions::invalid_request_exception(msg));
+        }
+        co_await remote().paxos_store().ensure_initialized(*schema);
```
You did not fix this one yet, right?

gusev-p commented on service/paxos/paxos_state.cc:
```diff
@@ -20,6 +20,7 @@
 #include "db/system_keyspace.hh"
```
Which fix? There are many commits that touch this file.

gusev-p commented on test/alternator/test_tablets.py:
```diff
@@ -104,11 +104,6 @@ def test_streams_enable_error_with_tablets(dynamodb):
-# (see #18068). We still allow creating an Alternator table with
-# tablets and using LWT for write isolation (always_use_lwt),
-# but the writes themselves will fail. When #18068 is fixed, this
-# test should be fixed to expect success - not failure.
```
I added the comment.

gusev-p commented on test/alternator/test_tablets.py:
```diff
@@ -118,14 +113,9 @@ def test_alternator_tablets_and_lwt(dynamodb):
-            table.put_item(Item={'p': 'hello'})
-            assert table.get_item(Key={'p': 'hello'})['Item'] == {'p': 'hello'}
+        table.put_item(Item={'p': 'hello'})
+        assert table.get_item(Key={'p': 'hello'})['Item'] == {'p': 'hello'}
```
done

gusev-p commented:
new version:
  * return back comments in alternator/test_tablets
  * handle the case when we're in RECOVERY mode, this means raft is disabled and schema changes are propagated through gossiper. We add support for this mode + a dedicated test.
  * `test_lwt`: add capital letters to test ks name and cf name to verify the use if quotes in CQL queries in `paxos_store`
  * reverted "paxos_state: read repair for intranode_migration" commit (it's part of a separate PR https://github.com/scylladb/scylladb/pull/24926, this PR is rebased on top of that one).
  * hide paxos state tables from "DESCRIBE SCHEMA": not decided yet, left as is for now (do not hide them), here's a discussion https://github.com/scylladb/scylladb/pull/24819#discussion_r2200321410
  * timeout for ensure_initialized + test
  * `create_write_response_handler/make_write_response_handler`, add a comment
  * rebranded `replica_local` flag to `node_local_only`, added reference to scylladb/scylladb#16478 in commit messages
  * moved LWT tests to `test_tablets_lwt.py`
  * explicit type constraint for `remove_non_local_host_ids`

avikivity commented on service/paxos/paxos_state.cc:
```diff
@@ -81,73 +81,6 @@ static dht::shard_replica_set shards_for_writes(const schema& s, dht::token toke
     return shards;
 }
 
-future<paxos_state> paxos_state::load_and_repair_paxos_state(db::system_keyspace& sys_ks,
```
> Any paxos state read which drives some paxos round forward is already
accompanied by a paxos state write. 

I guess this is the crux. If every read is also a write (to the same paxos row), then an obervation of the old state will heal the old+new state by overwriting it.

But it is always true? Can't we sometimes read and return failure (without writing anything new)?

gusev-p replied:
replied in https://github.com/scylladb/scylladb/pull/24926#issuecomment-3068522509

gusev-p replied:
yes, but #24926 is ready for merge

avikivity commented:
Please give this a soak test in S-C-T so the code paths have some runtime on them.

gleb-cloudius commented:
>     * handle the case when we're in RECOVERY mode, this means raft is disabled and schema changes are propagated through gossiper. We add support for this mode + a dedicated test.
> 

We do not support schema changes in the recovery mode IIRC. I think it is reasonable to not allow LWT operations in recovery mode if paxos table does not exist yet.

gleb-cloudius commented:
> > We do not support schema changes in the recovery mode IIRC. I think it is reasonable to not allow LWT operations in recovery mode if paxos table does not exist yet.
> 
> I have a test that proves otherwise, see `test_lwt_for_tablets_in_recovery_mode`. This test restarts two nodes in `RECOVERY` mode and verifies that LWT for tablets-based tables works, including the creation of the paxos state table. This didn't work before the latest fixes in this PR.

How writing a test proves anything? It is for us to decide if a scenario should be supported or not. Changing schema in recovery mode is unsupported. It means it is not tested and there may be code that assumes that it does not happen. If first LWT op requires schema change then it should not be allowed during recovery mode and return an error and the test should test for that.

gusev-p commented:
> How writing a test proves anything? It is for us to decide if a scenario should be supported or not. Changing schema in recovery mode is unsupported. It means it is not tested and there may be code that assumes that it does not happen. If first LWT op requires schema change then it should not be allowed during recovery mode and return an error and the test should test for that.

Discussed in Slack:
  * this recovery mode is not needed once the cluster is migrated to Raft-based topology.
  * we're working with tablet-based tables, so the cluster is actually migrated to tablets
  * no need to support the case when `migration_manager.use_raft()` returns false, we should just throw in both coordinator and replica

gusev-p commented:
> Question. How authentication is handled? Does paxos table need the same permissions as base table (probably yes). How is it handled?

I haven't thought about it, need to check.

timtimb0t commented:
> > Please give this a soak test in S-C-T so the code paths have some runtime on them.
> 
> The LWT dtests have passed with `--tablets` options ([link](https://github.com/scylladb/qa-tasks/issues/1918#issuecomment-3051649868)).
> 
> IIUC, currently there is no SCT workload which can verify LWT (@timtimb0t please correct if I'm wrong). We have a [separate task](https://github.com/scylladb/qa-tasks/issues/1920) to add such workload.

Yes, you are right. I'm preparing the SCT job to run LWT load

nyh reviewed (approved):
Thanks. I left a few comments and requests, but overall I'm marking "approve" because this is an impressive pull request and I think should go in perhaps after a couple of tiny changes I requested.
I'm not a Paxos expert, so I'm not sure that I understood all the finer details and pitfalls of the Paxos changes themselves, but I see that @gleb-cloudius already approved this PR, so great.

nyh commented on test/cluster/test_tablets2.py:
```diff
@@ -1957,3 +1958,54 @@ async def force_minor_compaction():
+
+            await cql.run_async(f"INSERT INTO {ks}.test (pk, c) VALUES ({i}, {i}) IF NOT EXISTS")
+            # For a Paxos round to succeed, a quorum of replicas must respond.
+            # Therefore, the Paxos state table must exist on at least a quorum of replicas,
```
I see. You're absolutely right - the driver doesn't know that an INSERT statement creates a table (!), so it doesn't wait for schema agreement.

nyh reviewed (approved):
Thanks. I'll merge.

If I understand correctly, the situation with Alternator after this patch is confusing. It now works well with tablets and LWT - but tablets are still NOT the default, right? Basically we'll need to do https://github.com/scylladb/scylladb/issues/22463. I'll do this as a followup.

gleb-cloudius commented:
I see thing like this:

```
ERROR 2025-07-20 13:38:56,660 [shard 0: gms] tablets - Base table acc19a40-6555-11f0-af53-5d174b8226c4 of co-located table acce1d60-6555-11f0-af53-5d174b8226c4 does not have a tablet map, at: 0x1581219 0x16dc240 0x1932b47 0x4983350 0x22d023c 0x22ced49 0x3d5aeca 0x1a8b54d 0x2acb478 0x2acabbf 0x2aca642 0x397d723 0x2acd802 /lib64/libc.so.6+0x35f4 /lib64/libc.so.6+0x36a7 0x31e3294
   --------
   seastar::internal::coroutine_traits_base<locator::tablet_metadata>::promise_type
   --------
   seastar::internal::coroutine_traits_base<void>::promise_type
   --------
   seastar::internal::coroutine_traits_base<void>::promise_type
   --------
   seastar::internal::coroutine_traits_base<void>::promise_type
   --------
   seastar::internal::coroutine_traits_base<void>::promise_type
   --------
   seastar::internal::coroutine_traits_base<void>::promise_type
   --------
   seastar::internal::coroutine_traits_base<void>::promise_type
   --------
   seastar::internal::coroutine_traits_base<void>::promise_type
Aborting on shard 0, in scheduling group gossip.
Backtrace:
  0x5842faa
  0x583ed9a
  0x58411ab
  0x586f94a
  /lib64/libc.so.6+0x19c2f
  /lib64/libc.so.6+0x7311b
  /lib64/libc.so.6+0x19afd
  /lib64/libc.so.6+0x16cf
  0x1932b9b
  0x4983350
  0x22d023c
  0x22ced49
  0x3d5aeca
  0x1a8b54d
  0x2acb478
  0x2acabbf
  0x2aca642
  0x397d723
  0x2acd802
  /lib64/libc.so.6+0x35f4
  /lib64/libc.so.6+0x36a7
  0x31e3294

```

Though strangely test.py does not create failed_test dir.

nyh commented:
Right, I see that too. E.g., https://jenkins.scylladb.com/job/scylla-master/job/next/9397/artifact/testlog/x86_64/dev/scylla-2774.log
```
INFO  2025-07-20 13:53:32,954 [shard 0: gms] schema_tables - Creating keyspace cql_test_1753008812952
INFO  2025-07-20 13:53:32,960 [shard 0: gms] migration_manager - Gossiping my schema version c758bf3a-6557-11f0-b4c0-c2a28001b8bf
INFO  2025-07-20 13:53:32,960 [shard 0: gms] schema_tables - Schema version changed to c758bf3a-6557-11f0-b4c0-c2a28001b8bf
WARN  2025-07-20 13:53:32,961 [shard 0:sl:d] create_keyspace - Using Replication Factor datacenter1=1 lower than the minimum_replication_factor_warn_threshold=3 is not recommended.
INFO  2025-07-20 13:53:32,962 [shard 0:comp] sstable - Rebuilding bloom filter /scylladir/testlog/x86_64/dev/scylla-2774/data/system/compaction_history-b4dbb7b4dc493fb5b3bfce6e434832ca/me-3gs4_0u98_5lwio2i9zxvwjotnt4-big-Filter.db: resizing bitset from 1448 bytes to 848 bytes. sstable origin: compaction
TRACE 2025-07-20 13:53:32,965 [shard 0: gms] group0_raft_sm - apply() is called with 1 commands
TRACE 2025-07-20 13:53:32,966 [shard 0: gms] group0_raft_sm - cmd: prev_state_id: optional(c758bf3a-6557-11f0-b4c0-c2a28001b8bf), new_state_id: c75aa20a-6557-11f0-1a0e-2822fa7c1588, creator_addr: 127.172.52.9, creator_id: 1c9dabc1-87f9-4f6b-be17-e91f8d67a7ae
TRACE 2025-07-20 13:53:32,966 [shard 0: gms] group0_raft_sm - cmd.history_append: {canonical_mutation: table_id 027e42f5-683a-3ed7-b404-a0100762063c schema_version fb2d07e3-5771-3937-955b-dd6afbd24917 partition_key pk{0007686973746f7279} partition_tombstone {tombstone: none}, row tombstone {range_tombstone: start={position: clustered, ckp{00109e76620a5fd711f08080808080808080}, 1}, end={position: clustered, ckp{}, 1}, {tombstone: timestamp=1753008812965121, deletion_time=1753008812}}{row {position: clustered, ckp{0010c75aa20a655711f01a0e2822fa7c1588}, 0} tombstone {row_tombstone: none} marker {row_marker: 1753008812965121 0 0}, column description atomic_cell{CQL DDL statement: "ALTER KEYSPACE cql_test_1753008812952 WITH replication = {'class' : 'NetworkTopologyStrategy', 'datacenter1' : 1 }",ts=1753008812965121,expiry=-1,ttl=0}}}
DEBUG 2025-07-20 13:53:32,966 [shard 0: gms] raft_topology - reload raft topology state
ERROR 2025-07-20 13:53:32,973 [shard 0: gms] tablets - Base table be7abbc0-6557-11f0-a4e7-26f3031f21c8 of co-located table be851c00-6557-11f0-a4e7-26f3031f21c8 does not have a tablet map, at: /jenkins/workspace/scylla-master/next/scylla/build/dev/seastar/libseastar.so+0x58338c /jenkins/workspace/scylla-master/next/scylla/build/dev/seastar/libseastar.so+0x583690 /jenkins/workspace/scylla-master/next/scylla/build/dev/seastar/libseastar.so+0x583998 /jenkins/workspace/scylla-master/next/scylla/build/dev/seastar/libseastar.so+0x3b9707 0x29c0a66 0x180c09b 0x1815da8 0x181e83a /jenkins/workspace/scylla-master/next/scylla/build/dev/seastar/libseastar.so+0x406c03 /jenkins/workspace/scylla-master/next/scylla/build/dev/seastar/libseastar.so+0x407f4a /jenkins/workspace/scylla-master/next/scylla/build/dev/seastar/libseastar.so+0x4092a7 /jenkins/workspace/scylla-master/next/scylla/build/dev/seastar/libseastar.so+0x408568 /jenkins/workspace/scylla-master/next/scylla/build/dev/seastar/libseastar.so+0x3797e7 /jenkins/workspace/scylla-master/next/scylla/build/dev/seastar/libseastar.so+0x378ad9 0x1315f31 0x13147df /lib64/libc.so.6+0x35f4 /lib64/libc.so.6+0x36a7 0x1313164
```

How don't know what these failures mean but they seem related to this patch (which indeed adds colocated tables). How did we not see these failures in CI? :-(

nyh commented:
Dequeued this patch from next :-(
@gusev-p can you please try to rebase your PR and have CI run again today, and/or think how come the PR passed with flying colors on CI but failed in next promotion?

nyh commented:
By the way, @xtrey, the faliures in https://jenkins.scylladb.com/job/scylla-master/job/next/9397/#showFailuresLink = fpr example https://jenkins.scylladb.com/job/scylla-master/job/next/9397/testReport/junit/cqlpy/test_keyspace/x86___release___test_py_release_x86___test_alter_keyspace_missing_rf_release_1/ -  are back to the situation were when Scylla crashes during one test file, instead of saying which test crashed, we see the next test file running on the same Scylla and failing on some silly attempt to connect to Scylla to do some setup like creating the test keyspace :-(

mlitvk commented on service/paxos/paxos_state.cc:
```diff
@@ -270,14 +278,386 @@ future<> paxos_state::learn(storage_proxy& sp, db::system_keyspace& sys_ks, sche
+    return container().invoke_on(0, &paxos_store::create_paxos_state_table, std::ref(s), timeout);
+}
+
+void paxos_store::on_before_drop_column_family(const schema& schema, utils::chunked_vector<mutation>& mutations, api::timestamp_type timestamp) {
+    if (const auto state_schema = try_get_paxos_state_schema(schema); state_schema) {
+        auto& keyspace = _db.find_keyspace(schema.ks_name());
+        auto muts = db::schema_tables::make_drop_table_mutations(keyspace.metadata(), state_schema, timestamp);
+        mutations.insert(mutations.end(), std::make_move_iterator(muts.begin()), std::make_move_iterator(muts.end()));
+    }
```
regarding the "Base table ... of co-located table ... does not have a tablet map" issue.
I think the problem is that the tablet map of the paxos table is not removed when the table is dropped
you need to notify here `before_drop_column_family` for the tablet allocator to remove the tablet map

gusev-p commented:
> Thanks. I'll merge.
> 
> If I understand correctly, the situation with Alternator after this patch is confusing. It now works well with tablets and LWT - but tablets are still NOT the default, right? Basically we'll need to do #22463. I'll do this as a followup.

There is a warning message, which is written to Scylla logs when  a new tablets-based keyspace is created:

> Tables in this keyspace will be replicated using Tablets and will not support Materialized Views, Secondary Indexes, CDC, LWT and counters features. To use Materialized Views, Secondary Indexes, CDC, LWT or counters, drop this keyspace and re-create it without tablets by adding AND TABLETS = {'enabled': false} to the CREATE KEYSPACE statement.

I'll remove 'LWT' from this list, but other features are still not ready, IIUC. I'm not sure do we need these features for Alternator.

gusev-p commented:
> @scylladb/scylla-maint - please review for inclusion.

This is actually on me right now -- I need to fix the failed tests. I got distracted by another issue.

enaydanov commented on test/pylib/scylla_cluster.py:
```diff
@@ -675,16 +675,25 @@ async def get_cql_up_state(self) -> ServerUpState | None:
         # Any other exception may indicate a problem, and is passed to the caller.
 
-    async def get_host_id(self, api: ScyllaRESTAPIClient) -> bool:
+    async def try_get_host_id(self, api: ScyllaRESTAPIClient) -> Optional[HostID]:
```
This change is not related to the PR and big enough to be a separate PR, imo.

gusev-p commented:
https://jenkins.scylladb.com/job/scylla-master/job/byo/job/byo_build_tests_dtest/3137 - `-m dtest_full` tests run

gusev-p commented:
> Dequeued this patch from next :-(
> @gusev-p can you please try to rebase your PR and have CI run again today, and/or think how come the PR passed with flying colors on CI but failed in next promotion?

The problem was that I forgot to remove the tablet map for paxos state tables. If `topology_state_load` is called after that, it can't find the base table (because it's already dropped) and crashes. The problem didn't reproduce in `test_tablets_lwt.test_lwt` despite it drops the table because of the hints parameter of the `topology_state_load` function. It's set most of the time, this means that only part of the system.tablets table is loaded. I added the `TRUNCATE TABLE` call to `test_tablets_lwt.test_lwt` which forces full `topology_state_load`.

Also I started a `-m dtest_full` tests run to check that all tests pass.

gusev-p commented:
> @gusev-p sorry for not being specific. The issue was not with dtest, but with `test.py`. The last status about `test.py` says all tests passed, but it was executing only changed tests and not full test set. BYO job was not affected, the bug was only for Scylla-CI pipeline that executes tests for PR's.

OK, but the BYO job runs all dtests and test.py tests, so we probably don't need to re-run Scylla-CI, right?

xtrey commented:
> > @gusev-p sorry for not being specific. The issue was not with dtest, but with `test.py`. The last status about `test.py` says all tests passed, but it was executing only changed tests and not full test set. BYO job was not affected, the bug was only for Scylla-CI pipeline that executes tests for PR's.
> 
> OK, but the BYO job runs all dtests and test.py tests, so we probably don't need to re-run Scylla-CI, right?

Partially, you're right. It runs all the test by default, and by default it runs only release mode. So your tests were executed only against release mode. Two options: rerun the Scylla-CI it will trigger all modes and unit custom tests with repeats for changed tests, run BYO with explicitly saying to run all modes.

gusev-p commented:
OK, restarted Scylla-CI https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18540/

nyh commented:
Re-merged. Hopefully second time is the charm :-)

piodul commented:
@gusev-p Sorry for being late to the party (and I only skimmed the commit titles), but I have one concern. The PR uses co-located tables for keeping the paxos state, right? I remember that you mentioned here: https://github.com/scylladb/scylladb/pull/22906#discussion_r2111431019 that the paxos tables should not be repaired. Right now, IIUC, when repairing the base table, all co-located tables are repaired. Does this PR opt-out the paxos table from repair somehow?

cc: @mlitvk
